### ISSUE TITLE
some doubts in the error part

### ISSUE THREAD

**Issue by akashprakas** [learnml-today/object-detection-with-pytorch#10]
As we have limited resources to answer an all questions, please read this rules before submit one.

### Rules:
1.    **Ask only questions related to course and not you personal projects, etc.** it doesn't mean that you can't ask them, but they will have lower priority and can be not answered at all.
2.    **Look on the Internet first.** Before submitting a question try to find answer on the Internet first, and ask question only if you didn't find the answer, or it not full, or you doesn't understand the answer.
3.    **Second, look on the closed issues.** You question can be already asked by someone else and get an answer.
4.    **No one know everything.** You should understand that ML area is very vast and we currently may have no answer for you. So it can take time to study your problem and prepare an answer.

**Comment by akashprakas** [comment]
Hi i have some doubts in the error function


responsible_bbox_arg = self.get_argmax_iou(pred, target) # [N,S,S]
this will be all ones right because of the argmax and argmax will return [1,1] for [-0.0,-0.0]
but eventually 
responsible_bbox_mask = responsible_bbox_mask * obj_mask # [N,S*S, B*5]

the above line makes everything work..can you please give a little more explanation here.

**Comment by creotiv** [comment]
So first of all when you have problems with understanding of transformations of matrices(we all have this problems cause it really hard to read and do modifications in mind) just brake them on parts and print out shapes and data on some small matrices.

Here is breakdown

```
responsible_bbox_arg = self.get_argmax_iou(pred, target) # [N,S,S]
print('responsible_bbox_arg',responsible_bbox_arg.shape)
print(responsible_bbox_arg)
# here we will get matrixs of tensors [0, 1, 2, 3, 4] for 1st box and 
# tensors [k*0, k*1, k*2, k*3, k*4] where k = ((self.B-1)*5)
responsible_bbox_scatter = torch.tensor((0, 1, 2, 3, 4))\
                        .repeat((num_batch, self.S * self.S, 1)).to(device)\
                        + responsible_bbox_arg.view(-1, self.S*self.S, 1)* ((self.B-1)*5) # [N,S*S, 5]
print('responsible_bbox_scatter',responsible_bbox_scatter)
responsible_bbox_mask = torch.zeros((num_batch, self.S * self.S, self.B * 5 + self.C)).to(device)\
                        .scatter_(2, responsible_bbox_scatter, torch.ones((num_batch, self.S * self.S, self.B * 5 + self.C)).to(device)) # [N,S*S, B*5]

print('responsible_bbox_mask',responsible_bbox_mask)
```

```
x = torch.rand((1,2,2,12))
y = torch.rand((1,2,2,12))
l = Loss()
l(x,y)
```

```
IOU tensor([[[[0.0265, -0.0000],
          [0.0142, -0.0000]],

         [[0.0000, 0.0447],
          [-0.0000, -0.0000]]]])
responsible_bbox_arg torch.Size([1, 2, 2])
tensor([[[0, 0],
         [1, 1]]])
responsible_bbox_scatter tensor([[[0, 1, 2, 3, 4],
         [0, 1, 2, 3, 4],
         [5, 6, 7, 8, 9],
         [5, 6, 7, 8, 9]]])
responsible_bbox_mask tensor([[[1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0.],
         [1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 0., 0.],
         [0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 0., 0.]]])
(tensor(15.6756),
 {'class_loss': 0.08544204384088516,
  'noobj_loss': 0.0,
  'obj_loss': 0.0479549765586853,
  'wh_loss': 7.192577362060547,
  'xy_loss': 8.349601745605469})
```

So responsible_bbox_arg returns the presence of an object. as we have only a present or not then it will be 0,1.

As our data encode in B*5+C then in the current example we will have 12 values, and we need to mask only those that have an object. So first box or second. and that params from 0 to 4 or from 5 to 9. That what responsible_bbox_scatter do.

Then responsible_bbox_scatter encode in binary mask responsible_bbox_mask and applied with obj_mask: responsible_bbox_mask = responsible_bbox_mask * obj_mask, which is boolean mask saying which target has object.

Then responsible_bbox_mask applied to loss to calculate the only loss for objects presence.

**Comment by akashprakas** [comment]
Thanks for the explanation. 
IOU tensor([[[[0.0265, -0.0000],
          [0.0142, -0.0000]],

         [[0.0000, 0.0447],
          [-0.0000, -0.0000]]]])

this is the iou tensor and last row is [-0.00,-0.00]

now in the responsible_bbox_arg the index returned for that is 1. But there are no objects there, so dn't we need another method,but it works fine as we multiply by object mask down.
Thanks for the explanation.

**Comment by creotiv** [comment]
@akashprakas Sorry, give you the wrong answer on your question, mixed up with another network.
So forget what i told before.

So here with an IoU we check not the presence of the object but selecting which of N boxes that we predict best matches with target boxes. So argmax just saying which of two boxes we should use to compute the loss. When we get IoU for both of them 0.0 it's mean that we no really care which to choose(but better, of course, would be to choose that one which have center coordinates nearer to our target box, so network needs to move fewer weights).

One more time sorry for being wrong and lead you to the wrong direction, it was a hard week and I was brainwashed a little

**Comment by creotiv** [comment]
But you also find an interesting bug in torch.argmax function that instead of returning first argmax object returns the last one, and also can be inconsistent in such scenarios when there can be same values in matrix.

**Comment by akashprakas** [comment]
thank you :)
